### ISSUE TITLE
Insufficient permissions error with Zone-specific API tokens

### ISSUE THREAD

**Issue by tornyairoland** [terraform-providers/terraform-provider-cloudflare#702]
Hi!

I was directed here from [this](https://github.com/pulumi/pulumi-cloudflare/issues/82) pulumi-cloudflare issue.

While trying to limit the access of our Cloudflare API Tokens, I have come across an error.
I have created a new Token, added permissions to "Edit" the "Zone Settings" and "Included" only that "Specific Zone".
![kép](https://user-images.githubusercontent.com/48458705/84165420-f8c76700-aa73-11ea-86f5-73eafaafec04.png)

### Terraform Version
The Pulumi Cloudflare provider is using the Cloudflare Terraform provider, v2.7.0, and ` github.com/terraform-providers/terraform-provider-cloudflare v1.18.2-0.20200520193353-7dbdc13b06b6`.

### Affected Resource(s)
I tried it with `cloudflare_zone_settings_override`, but other `cloudflare_*` resources could be affected.

### Terraform Configuration Files
Based on the pulumi code I tried to run:
```hcl
resource "cloudflare_zone_settings_override" "settings" {
    zone_id = var.cloudflare_zone_id
    settings {
        ssl = "flexible"
        min_tls_version = "1.2"
    }
}
```
I tried getting the cloudflare_zone_id programmatically and copying the Zone ID manually from Cloudflare as well.

### Expected Behavior
Zone settings should have been applied as the token used for this has the permissions to do that.

### Actual Behavior
Got the following error from Pulumi:
```
Error: invocation of cloudflare:index/getZones:getZones returned an error: invoking cloudflare:index/getZones:getZones: error listing Zone: error from makeRequest: HTTP status 403: insufficient permissions
```

### Steps to Reproduce
1. Create an API Token limited to just a specific Zone, nothing Account related.
2. Try to apply some setting to just the specified Zone.

### Important Factoids
1. Using the same token and calling the Cloudflare API with curl works:
```
$ curl -X GET "https://api.cloudflare.com/client/v4/zones?name=<zone name>" -H "Authorization: Bearer <api token>" -H "Content-Type:application/json"

{"result":[{"id":"be2[...]","name":"<zone name>","status":"active",[...]}
```
2. Using the same curl command, but without the `?name=<zone name>` part returns an error from the Cloudflare API:
```
{"success":false,"errors":[{"code":0,"message":"Actor 'com.cloudflare.api.token.81a[...]' requires permission 'com.cloudflare.api.account.zone.list' to list zones"}],"messages":[],"result":null}
```
3. Some related info is mentioned on the Cloudflare forums [here](https://community.cloudflare.com/t/bug-zone-detail-by-name-requires-zone-list-permission/128042/14)

### References
[Pulumi issue](https://github.com/pulumi/pulumi-cloudflare/issues/82)

**Comment by jacobbednarz** [comment]
Are you able to re-run the command with `TF_LOG=DEBUG` and we can see which HTTP call is getting this exception?

**Comment by jacobbednarz** [comment]
> The Pulumi Cloudflare provider is using the Cloudflare Terraform provider, v2.7.0, and github.com/terraform-providers/terraform-provider-cloudflare v1.18.2-0.20200520193353-7dbdc13b06b6.

Would you also be able to expand on this as the way I read this, it's conflicting information. You can't be running two different versions of the provider (2.7.0 and 1.18.2) at the same time.

**Comment by mrtristan** [comment]
i'm seeing much the same thing. my api tokens have zone specific permissions. 

when doing:
```
data "cloudflare_zones" "XXXXXXX" {
  filter {
    name   = "xxxx.com"
    # status = "active"
    # paused = false
  }
}
```

fiddler captured:
```
GET https://api.cloudflare.com/client/v4/zones?per_page=50 HTTP/1.1
Host: api.cloudflare.com
User-Agent: HashiCorp Terraform/0.12.26 (+https://www.terraform.io) Terraform Plugin SDK/1.13.0 terraform-provider-cloudflare/2.7.0
Authorization: Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
Content-Type: application/json
Accept-Encoding: gzip
```

it seems that the api calls are too generic when the api token only has a subset of access.

a curl like this works with the same token:
```
curl -X GET "https://api.cloudflare.com/client/v4/zones?name=xxxx.com" -H "Authorization: Bearer XXXXXXXXXXXXXXXXXXX" -H "Content-Type:application/json"
```

is this sufficiently similar? not looking to cause duplication, and not looking to hijack... :-)

**Comment by jacobbednarz** [comment]
@mrtristan Your issue could be similar but really should be a separate issue as it's not related to Pulumi from my understanding here. With your API calls, it's failing because we do the [filtering in the application](https://github.com/terraform-providers/terraform-provider-cloudflare/blob/master/cloudflare/data_source_zones.go#L68) following fetching all the zones which is probably why your seeing the issue. We can probably look to do the filtering on Cloudflare's side to mitigate but we'll need to do some testing that we don't lose any functionality in doing so.

**Comment by mrtristan** [comment]
yep. completely aligned. moving to new issue

**Comment by gstotts** [comment]
I seem to be having a similar issue although I'm not using Zone specific api keys.  I use the global key for an account.  When we went to update from terraform 11 to terraform 12 but using provider 2.7.0, W e made the adjustments to our terraform but are now getting a non-descript insufficient privileges error.   I'm not sure if it's a state issue (calling something that doesn't exist?) or access but the behavior is the same as what hey describes here.

Running debug only provided me the same -- I show a ton of calls to Cloudflare API that get 200s prior to that specific error.  I reached out to Cloudflare support and they wanted me to go here.  Any suggestion would be helpful.  Thanks.

**Comment by tornyairoland** [comment]
Hi!
I recreated the issue using terraform and not pulumi, and the same error is happening.
I think it is similar to the issue that @mrtristan mentioned.

Cloudflare Terraform provider: **v2.8.0**  (the other version number above was the go package version from the provider repository)

Here are the debug logs: https://gist.github.com/tornyairoland/6cd8ab4146851b97273545fa114af671

**Comment by jacobbednarz** [comment]
@tornyairoland the issue @mrtristan raised was moved to #707 and the fix is currently under review at #708. If you believe you're having the same issue, please follow that issue and checkout the PR if you're wanting to test out the fix.

**Comment by jacobbednarz** [comment]
#708 has been merged so fetching a single zone shouldn't be causing these issues (providing Cloudflare permissions are correct).

Feel free to pull the `master` branch down, build it per the README documentation and let me know if you hit any snags. Alternatively, this will land in the next release.
